### ISSUE TITLE
indexer support for A RDATA in DSC collector

### ISSUE THREAD

**Issue by hassaniqbal2k3** [DNS-OARC/dsc#6]
Need to capture RDATA for TYPE A from the DNS response message. Currently DSC collector does not have support for this.

For A RDATA format I am referring to RFC at - https://www.ietf.org/rfc/rfc1035.txt

3.4.1. A RDATA format

    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+
    |                    ADDRESS                    |
    +--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+--+

where:

ADDRESS         A 32 bit Internet address.

**Comment by jelu** [comment]
I think the purpose of DSC (DNS **Stats** Collector) is to gather statistical information about the DNS packages. To implement something like this is kinda out of scope and currently the RDATA is not processed at all so its not a simple issue of adding another indexer.

What do you say @wessels ?

**Comment by wessels** [comment]
Agreed that DSC might not be the best tool this.  Currently it does not parse any RDATA, although it certainly could in theory.  

There would be some complexity in doing so.  For example a response might have multiple A RRs (do you count all of them or just one?)  The response might include out-of-bailiwick RRs that would be ignored by a proper resolver (do you count them or not?).  Do you follow CNAME/DNAMEs or not?

**Comment by jelu** [comment]
@hassaniqbal2k3 Can you describe why you want to extract A RDATA?

**Comment by hassaniqbal2k3** [comment]
@jelu To build stats for an RBLDNS system. We have found DSC to be the closet tool out there that we can leverage. RDATA is important for this. I have added hook to the dsc source to get A RDATA. Now i am working to build indexer. Once done, will create a fork on git to get your feedback.

**Comment by hassaniqbal2k3** [comment]
Folks i could move one more step on this. i am able to add the indexer for A RDATA. XML file is getting generated nicely for one A record. But the challenge that i am facing now is to handle mutiple A RRs, as mentioned by @wessels. i have added inX_addr ardata_ip_addr in dns_message struct. For multiple A records this is getting over written in the code that is there for dns packet processing as of now. First step towards handling the multiple A RRs, i thought of making it a linked list and loop over it in ardata_indexer (newly added). But looks like that may not solve the problem, as to me it appears that indexing is happening only once per dns packet. 

@wessels/ @jelu - can you please give some clue what file/function should i be looking to handle this. I have not got much success to nail the code where dns_message struct is getting utilized to pass it on to various indexers to index the data. That section seems to be the potential area to add this support. I will continue digging more into DSC source in the mean time. Any clue is highly appreciated.

**Comment by jelu** [comment]
I seem to have received a comment over email but looks like it has been removed from this issue.

@hassaniqbal2k3 if you have any further questions I think it would be a good idea to upload the changes you have made first so that we can look at them. You do not have to make a pull request, just a commit and push to your fork.

**Comment by hassaniqbal2k3** [comment]
@jelu i have made code changes to support the A RDATA indexer and it seems to work for single A  RR, but incase of multiple A RRs the stats were not adding up and i figured out what was going on. But then i had question regarding areas where i can make changes for handling multiple A RR case. Then on spending some more time with code i figured out the flow where this change can be hacked in. So to save everybody's time i removed the question from here. it was pretty long question as well and had potential to cause confusion.

I have made changes for handling multiple A RR case and testing is in progress. I will  commit and push the code to my fork for reviews once i am done with testing at my end. 

Thank you for the follow up!

**Comment by jelu** [comment]
When you have workable code you can make a pull request for a review.

**Comment by hassaniqbal2k3** [comment]
Sure. we have moved quite a bit with dsc collector changes. let me sync that in one by one. i will create a pull request at the earliest opportunity.
